crates/transport_tx5/src/test.rs

```diff
@@ -525,9 +536,161 @@ async fn preflight_send_recv() {
+    })
+    .await
+    .unwrap();
+
+    assert!(Some(faulty_url) == r);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn offline_peer_marked_unresponsive() {
+    // set the tx5 timeout to 5 seconds to keep the test reasonably short
+    let test = Test::new(None, Some(5)).await;
+
+    let (unresp_send, mut unresp_recv) = tokio::sync::mpsc::unbounded_channel();
+
+    let u1 = Arc::new(Mutex::new(Url::from_str("ws://bla.bla:38/1").unwrap()));
+    let u1_2 = u1.clone();
+    let h1 = Arc::new(CbHandler {
+        new_addr: Arc::new(move |url| {
+            *u1_2.lock().unwrap() = url;
+        }),
+        mark_peer_unresponsive: Arc::new({
+            move |url| {
+                unresp_send.send(url).map_err(|_| K2Error::Other {
+                    ctx: "Failed to send url to oneshot channel".into(),
+                    src: DynInnerError::default(),
+                })
+            }
+        }),
+        ..Default::default()
+    });
+    let t1 = test.build_transport(h1.clone()).await;
+    t1.register_space_handler(TEST_SPACE_ID, h1.clone());
+    t1.register_module_handler(TEST_SPACE_ID, "mod".into(), h1.clone());
+
+    let (s_send, mut s_recv) = tokio::sync::mpsc::unbounded_channel();
+    let u2 = Arc::new(Mutex::new(Url::from_str("ws://bla.bla:38/1").unwrap()));
+    let u2_2 = u2.clone();
+    let h2 = Arc::new(CbHandler {
+        new_addr: Arc::new(move |url| {
+            *u2_2.lock().unwrap() = url;
+        }),
+        recv_space_notify: Arc::new(move |url, space, data| {
+            let _ = s_send.send((url, space, data));
+            Ok(())
+        }),
+        ..Default::default()
+    });
+    let t2 = test.build_transport(h2.clone()).await;
+    t2.register_space_handler(TEST_SPACE_ID, h2.clone());
+    t2.register_module_handler(TEST_SPACE_ID, "mod".into(), h2.clone());
+
+    let u2: Url = u2.lock().unwrap().clone();
+    println!("got u2: {}", u2);
+
+    // check that send works initially while peer 2 is still online
+    t1.send_space_notify(
+        u2.clone(),
+        TEST_SPACE_ID,
+        bytes::Bytes::from_static(b"hello"),
+    )
+    .await
+    .unwrap();
+
+    // checks that recv works
+    let r = tokio::time::timeout(std::time::Duration::from_secs(5), async {
+        s_recv.recv().await
+    })
+    .await
+    .unwrap()
+    .unwrap();
+
+    println!("{r:?}");
+    assert_eq!(b"hello", r.2.as_ref());
+
+    // Check that the peer has not been marked as unresponsive
+    let r = tokio::time::timeout(std::time::Duration::from_secs(3), async {
+        unresp_recv.recv().await
+    })
+    .await;
+
+    assert!(r.is_err());
+
+    // Now peer 2 goes offline and we check that it gets marked as unresponsive
+    drop(t2);
+
+    // We need to wait for a while in order for the webrtc connection to get
+    // disconnected. If this test becomes flaky, this waiting period may need
+    // to be increased a bit.
+    tokio::time::sleep(std::time::Duration::from_secs(7)).await;
```

Comment:
This is a bit unfortunate but due to the nature of webrtc, the connection will only close 6-7 seconds after the peer goes offline so any send attempt earlier than that will look like it succeeded from the sender side. Hence why we need to wait a bit here.

Reply:
That's interesting. So "not getting a response" isn't enough. We actually do need the connection to be come invalid...

Okay, well we can come back and look for other ways to force the connection to break faster if we have problems here. It is a bit slow then this test but it's an important thing to test.

I am wondering if we should consider moving slower tests like this to the integration tests so that `cargo test --lib` stays fast. But then we probably don't want to do that on this PR because the test harness will have to be exposed from the crate to be used in integration tests.